picard/AddOrReplaceReadGroups.wdl

```diff
@@ -0,0 +1,86 @@
+version 1.0
+# -------------------------------------------------------------------------------------------------
+# Package Name: https://broadinstitute.github.io/picard/
+# Task Summary: Add/replace read groups in a BAM file.
+# Tool Name: Picard AddOrReplaceReadGroups
+# Documentation: https://broadinstitute.github.io/picard/command-line-overview.html#AddOrReplaceReadGroups
+# -------------------------------------------------------------------------------------------------
+
+
+task AddOrReplaceReadGroups {
+  input {
+    File ? java
+    File ? picard
+
+    File input_bam
+    String sample_id
+    String output_filename_extension
+
+    Int RGID = 4
+    String RGLB = "lib1"
+    String RGPL = "illumina"
+    String RGPU = "unit1"
+    String RGSM = "20"
+
+    String ? userString
+
+    Array[String] modules = []
+    Int memory = 4
+    Int cpu = 1
+  }
+
+  String output_filename = "${sample_id}" + "${output_filename_extension}"
```

Comment:
Is there a reason why it need to be a particular extension? This generally would be an intermediate file and Cromwell would store it in it's own working directory.  You can then pass this output to other tasks like: `AddOrReplaceReadGroups.output_bam`

Reply:
This is mainly an attempt to replicate what is seen in the legacy Pegasus pipeline, where the `input_bam` has one extension and the output file will have a different extension in order to identify what has changed (think `Aligned.bam` vs. `Aligned.rg.bam`). My first thought was to just have `Aligned.rg.bam` passed in the input template as `...AddOrReplaceReadGroups.output_filename`, but then we lose the sample ID prefix. 

This would not be an issue if this task wasn't the last one to touch the bam file, but it is - the tasks downstream do not output the bam that is fed into them by this task. But I do recognize the strong bias this has for the RNA-Seq pipeline...

There are two options:
1. The input template for RNA-Seq just hard codes the final bam filename with the sample ID, e.g. `test002.Aligned.rg.bam` or...
2. Drop the sample ID all together and just use a tasks like `FinalizeOutputs` to do all the housekeeping at the end

Either way, I see that this needs to be less pipeline-specific

Reply:
I think this is elegant the way you have it.  I was thinking of something like this: `String output_filename = sample_id + ".rg.bam"`.  I didn't think there was a need to set/parameterize the output file extension.

Reply:
That might be sufficient but I think this way provides a little more control. Only problem is, it's not intuitive and would require some investigation if this were to be used elsewhere. So I could implement what you have if we want to make this as pipeline-agnostic as possible.

Reply:
yea or just provide a default "rg.bam" - so those who don't want to set it don't have to

Reply:
I like it, see latest commit

Reply:
👍